pkg/common/resources/resources.go

```diff
@@ -44,6 +46,52 @@ func (q Quantity) string() string {
 	return strconv.FormatInt(int64(q), 10)
 }
 
+// Util struct to keep track of application resource usage
+type UsedResourceTracker struct {
+	// Two level map for aggregated resource usage
+	// With instance type being the top level key, the mapped value is a map:
+	//   resource type (CPU, memory etc) -> the aggregated used time (in seconds) of the resource type
+	//
+	UsedResourceMap map[string]map[string]int64
+
+	sync.RWMutex
+}
+
+// Aggregate the resource usage to UsedResourceMap[instType]
+// The time the given resource used is the delta between the resource createTime and currentTime
+func (urt *UsedResourceTracker) AggregateUsedResource(instType string,
+	resource *Resource, createTime time.Time) {
+	urt.Lock()
+	defer urt.Unlock()
+
+	releaseTime := time.Now()
+	timeDiff := int64(releaseTime.Sub(createTime).Seconds())
+	aggregatedResourceTime, ok := urt.UsedResourceMap[instType]
+	if !ok {
+		aggregatedResourceTime = map[string]int64{}
+	}
+	for key, element := range resource.Resources {
+		resourceSecondsKey := key + "Seconds"
+		curUsage, ok := aggregatedResourceTime[resourceSecondsKey]
+		if !ok {
+			curUsage = 0
+		}
+		curUsage += int64(element) * timeDiff  // resource size times timeDiff
+		aggregatedResourceTime[resourceSecondsKey] = curUsage
+	}
+	urt.UsedResourceMap[instType] = aggregatedResourceTime
+}
+
+func (urt *UsedResourceTracker) GetResourceUsageSummary() string {
+	urt.Lock()
+	defer urt.Unlock()
+	urm, err := json.Marshal(urt.UsedResourceMap)
```

Comment:
We should not be marshalling into JSON here, as the zap logger can do that for us. Instead, we should log individual fields.

Reply:
HI Craig, that means we need to create a snapshot (deep copy) of the UsedResourceMap and return this snapshot here instead of the marshalled string, and then have zap logger report the map as json.  That's what you suggested right?

 @wilfred-s raised the concern that map uses more memory than one would expect earlier, and I tested it, it's indeed a lot of more memory consumption than I expected. Plus the runtime to do deep copy. I wonder if we really want to go this route now, or we can defer to later when there is a consumer requiring map instead of string? This is internal to yunikorn-core right now and the app summary is only used for the logging added by this ticket. 

Thanks.

Reply:
This also needs to be able to be turned off via zap logger levels. Right now, every user will pay the cost of marshalling this complex object every time, even if logging of these events is disabled. That's part of what the zap logger framework gets you. It also standardizes representation of objects in the log so that the log formats are consistent throughout the application. I'll say again: This MUST not be a string returned here.

Reply:
HI @craigcondit , wonder if this would work:

I renamed UsedResourceTracker to UsedResource here, and added Clone() method in UsedResource,

func (ur *UsedResource) Clone() *UsedResource {
	if ur == nil:
		return nil

	ret := NewUsedResource()
	ur.RLock()
	defer ur.RUnlock()
	for k, v := range ur.UsedResourceMap {
		sourceEntry := map[string]int64(v)
		destEntry := make(map[string]int64)
		for key, element := range sourceEntry {
			destEntry[key] = element
		}
		ret.UsedResourceMap[k] = destEntry
	}
	return ret
}

func (sa *Application) GetApplicationSummary(rmID string) *ApplicationSummary {
	sa.RLock()
	defer sa.RUnlock()
	appSummary := &ApplicationSummary{
		ApplicationID: sa.ApplicationID,
		SubmissionTime: sa.SubmissionTime,
		StartTime: sa.startTime,
		FinishTime: sa.finishedTime,
		User: sa.user.User,
		Queue: sa.queuePath,
		State: sa.stateMachine.Current(),
		RmID: rmID,
		**ResourceUsage: sa.usedResource.Clone(),**
	}
	return appSummary
}

func (as *ApplicationSummary) GetReport() string {
	res := "{"
	res += "\"appId\":\"" + as.ApplicationID + "\","
	res += "\"submissionTime\":\"" + strconv.FormatInt(as.SubmissionTime.UnixMilli(), 10) + "\","
	res += "\"startTime\":\"" + strconv.FormatInt(as.StartTime.UnixMilli(), 10) + "\","
	res += "\"finishTime\":\"" + strconv.FormatInt(as.FinishTime.UnixMilli(), 10) + "\","
	res += "\"user\":\"" + as.User + "\","
	res += "\"queue\":\"" + as.Queue + "\","
	res += "\"state\":\"" + as.State + "\","
	res += "\"rmID\":\"" + as.RmID + "\","
	**res += "\"resourceUsage\":" + as.ResourceUsage.GetResourceUsageJson()**
	res += "}"
	return res
}
Above I introduced a GetResourceUsageJson() because I was doing different processing with the different fields you can see in the above method, which abstract what information we need to report in what format.

func (sa *Application) GetAppSummary(rmID string) string {
	appSummary := sa.GetApplicationSummary(rmID)
	ret := appSummary.GetReport()
	appSummary.ResourceUsage = nil
	return ret

And the final report is created with:

	log.logger().Info(
		zap.String("YK_APP_SUMMARY", app.GetAppSummary(pc.RmID)))

inside  func (pc *PartitionContext) moveTerminatedApp(appID string) {

If user chooses to use log level higher than info, then, s/he may not get the report. 

To be honest, we actually hope we can always report this without being impacted by the log level.

Thanks

Reply:
This is not YARN, we don't do the same things. We have to play nice within the Kubernetes ecosystem, where logging all goes through a single output (stdout). Because of this, we MUST be consistent in how we log, or we make it difficult for ALL downstream consumers of log events. We can't just change things up for the convenience of your consumer code. 

As for this:

`log.logger().Info(zap.String("YK_APP_SUMMARY", app.GetAppSummary(pc.RmID)))`

You will end up evaluating `app.GetSummary(pc.RmID)` whether or not logging is enabled. To avoid this, one method would be to use `Stringer()` and provide a deferred function that retrieves the app summary as a string. However, this breaks the structured logging as you are logging a complex object. It's better to log this as individual fields so that the zap log line can be parsed appropriately.